src/backend/distributed/connection/placement_connection.c

```diff
@@ -723,38 +722,83 @@ ResetPlacementConnectionManagement(void)
- * the entire coordinated transaction.
+ * MarkFailedShardPlacements looks through every connection in the connection shard hash
+ * and marks the placements associated with failed connections invalid.
  *
- * This will usually be called twice. Once before the remote commit is done,
- * and once after. This is so we can abort before executing remote commits,
- * and so we can handle remote transactions that failed during commit.
+ * Every shard must have at least one placement connection which did not fail. If all
+ * modifying connections for a shard failed then the transaction will be aborted.
  *
- * When preCommit or using2PC is true, failures on transactions marked as
- * critical will abort the entire coordinated transaction. If not we can't
- * roll back, because some remote transactions might have already committed.
+ * This will be called just before commit, so we can abort before executing remote
+ * commits. It should also be called after modification statements, to ensure that we
+ * don't run future statements against placements which are not up to date.
  */
 void
-CheckForFailedPlacements(bool preCommit, bool using2PC)
+MarkFailedShardPlacements()
+{
+	HASH_SEQ_STATUS status;
+	ConnectionShardHashEntry *shardEntry = NULL;
+
+	hash_seq_init(&status, ConnectionShardHash);
+	while ((shardEntry = (ConnectionShardHashEntry *) hash_seq_search(&status)) != 0)
+	{
+		if (!CheckShardPlacements(shardEntry))
+		{
+			ereport(ERROR,
+					(errmsg("was not able to make changes to shard " INT64_FORMAT
+							" on any node",
+							shardEntry->key.shardId)));
+		}
+	}
+}
+
+
+/*
+ * PostCommitMarkFailedShardPlacements marks placements invalid and checks whether
+ * sufficiently many placements have failed to abort the entire coordinated
+ * transaction.
+ *
+ * This will be called just after a coordinated commit so we can handle remote
+ * transactions which failed during commit.
+ *
+ * When using2PC is set as least one placement must succeed per shard. If all placements
+ * fail for a shard the entire transaction is aborted. If using2PC is not set then a only
+ * a warning will be emitted; we cannot abort because some remote transactions might have
+ * already been committed.
+ */
+void
+PostCommitMarkFailedShardPlacements(bool using2PC)
 {
 	HASH_SEQ_STATUS status;
 	ConnectionShardHashEntry *shardEntry = NULL;
 	int successes = 0;
 	int attempts = 0;
 
+	int elevel = using2PC ? ERROR : WARNING;
+
 	hash_seq_init(&status, ConnectionShardHash);
 	while ((shardEntry = (ConnectionShardHashEntry *) hash_seq_search(&status)) != 0)
 	{
 		attempts++;
-		if (CheckShardPlacements(shardEntry, preCommit, using2PC))
+		if (CheckShardPlacements(shardEntry))
 		{
 			successes++;
 		}
+		else
+		{
+			/*
+			 * Only error out if we're using 2PC. If we're not using 2PC we can't error
+			 * out otherwise we can end up with a state where some shard modifications
+			 * have already committed successfully.
+			 */
+			ereport(elevel,
```

Comment:
I'm not entirely sure why we would error out here when using 2PC. The transaction succeeded, we just didn't manage to commit everything, but we should be able to do that later.